clang/include/clang/AST/PreorderAST.h

```diff
@@ -0,0 +1,124 @@
+    Node *Parent, *Left, *Right;
+
+    Node(Node *Parent) :
+      Opc(BO_Add), HasConst(false),
+      Parent(Parent), Left(nullptr), Right(nullptr) {}
+
+    // Is the operator commutative and associative?
+    bool IsOpCommutativeAndAssociative() {
+      return Opc == BO_Add || Opc == BO_Mul;
+    }
+  };
+
+  class PreorderAST {
+  private:
+    ASTContext &Ctx;
+    Lexicographic Lex;
+    llvm::raw_ostream &OS;
+    bool Error;
+    Node *Root;
+
+    // Create a PreorderAST for the expression E.
+    // @param[in] E is the sub expression which needs to be added to N.
+    // @param[in] N is the current node of the AST.
+    // @param[in] Parent is the parent node for N.
+    void Create(Expr *E, Node *N = nullptr, Node *Parent = nullptr);
+
+    // Sort the variables in a node of the AST.
+    // @param[in] N is current node of the AST.
+    void Sort(Node *N);
+
+    // Check if the two AST nodes N1 and N2 are equal.
+    // @param[in] N1 is the first node.
+    // @param[in] N2 is the second node.
+    // @return Returns a boolean indicating whether N1 and N2 are equal.
+    bool IsEqual(Node *N1, Node *N2);
+
+    // Set Error in case an error occurs during transformation of the AST.
+    void SetError() { Error = true; }
+
+    // Print the PreorderAST.
+    // @param[in] N is the current node of the AST.
+    void PrettyPrint(Node *N);
+
+    // Cleanup the memory consumed by node N.
+    // @param[in] N is the current node of the AST.
+    void Cleanup(Node *N);
+
+    // A DeclRefExpr can be a reference either to an array subscript (in which
+    // case it is wrapped around a ArrayToPointerDecay cast) or to a pointer
+    // dereference (in which case it is wrapped around an LValueToRValue cast).
+    // @param[in] An expression E.
+    // @return Returns a DeclRefExpr if E is a DeclRefExpr, otherwise nullptr.
+    DeclRefExpr *GetDeclOperand(Expr *E);
+
+  public:
+    PreorderAST(ASTContext &Ctx, Expr *E) :
+      Ctx(Ctx), Lex(Lexicographic(Ctx, nullptr)), OS(llvm::outs()),
+      Error(false), Root(nullptr) {
+      Create(E);
+    }
+
+    // Normalize the input expression through a series of transforms on the
+    // preorder AST. The Error field is set if an error is encountered during
+    // transformation of the AST.
+    void Normalize();
+
+    // Check if the two ASTs are equal. This is intended to be called from
+    // outside this class and invokes IsEqual on the root nodes of the two ASTs
+    // to recursively compare the AST nodes.
+    // @param[in] this is the first AST.
+    // @param[in] P is the second AST.
+    // @return Returns a bool indicating whether the two ASTs are equal.
+    bool IsEqual(PreorderAST &P) { return IsEqual(Root, P.Root); }
+
+    // Check if an error has occurred during transformation of the AST.
+    // @return Whether an error has occurred or not.
+    bool GetError() { return Error; }
+
+    // Cleanup the memory consumed by the AST. This is intended to be called
```

Comment:
This is a public method: do you need to mention that it is intended to be called from outside this class?